note/note_rfc6962.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	logID := sha256.Sum256(key)
+	v.keyHash = rfc6962Keyhash(name, logID)
+	v.v = verifyRFC6962(pubK)
+
+	return v, nil
+}
+
+// SignedTreeHead represents the structure returned by the get-sth CT method
+// after base64 decoding; see sections 3.5 and 4.3.
+type signedTreeHead struct {
+	Version           int    `json:"sth_version"`         // The version of the protocol to which the STH conforms
+	TreeSize          uint64 `json:"tree_size"`           // The number of entries in the new tree
+	Timestamp         uint64 `json:"timestamp"`           // The time at which the STH was created
+	SHA256RootHash    []byte `json:"sha256_root_hash"`    // The root hash of the log's Merkle tree
+	TreeHeadSignature []byte `json:"tree_head_signature"` // Log's signature over a TLS-encoded TreeHeadSignature
+	LogID             []byte `json:"log_id"`              // The SHA256 hash of the log's public key
+}
+
+// RFC6962STHToCheckpoint converts the provided RFC6962 JSON representation of a SignedTreeHead structure to
+// a sunlight style signed checkpoint.
+// The passed in verifier must be an RFC6929Verifier containing the correct details for the log which signed the STH.
+func RFC6962STHToCheckpoint(j []byte, v note.Verifier) ([]byte, error) {
+	var sth signedTreeHead
+	if err := json.Unmarshal(j, &sth); err != nil {
+		return nil, err
+	}
+	logName := v.Name()
+	body := fmt.Sprintf("%s\n%d\n%s\n", logName, sth.TreeSize, base64.StdEncoding.EncodeToString(sth.SHA256RootHash))
+	sigBytes := binary.BigEndian.AppendUint32(nil, v.KeyHash())
+	sigBytes = binary.BigEndian.AppendUint64(sigBytes, sth.Timestamp)
+	sigBytes = append(sigBytes, sth.TreeHeadSignature...)
+	sigLine := fmt.Sprintf("\u2014 %s %s", logName, base64.StdEncoding.EncodeToString(sigBytes))
+
+	n := []byte(fmt.Sprintf("%s\n%s\n", body, sigLine))
+	if _, err := note.Open(n, note.VerifierList(v)); err != nil {
+		return nil, err
+	}
+	return n, nil
+}
+
+// RFC6962STHTimestamp extracts the embedded timestamp from a translated RFC6962 STH signature.
+func RFC6962STHTimestamp(s note.Signature) (time.Time, error) {
+	r, err := base64.StdEncoding.DecodeString(s.Base64)
+	if err != nil {
+		return time.UnixMilli(0), errMalformedSig
+	}
+	if len(r) <= keyHashSize+timestampSize {
+		return time.UnixMilli(0), errVerifierAlg
+	}
+	r = r[keyHashSize:] // Skip the hash
+	// Next 8 bytes are the timestamp as Unix millis-since-epoch:
+	return time.Unix(0, int64(binary.BigEndian.Uint64(r)*1000)), nil
+}
+
+func rfc6962Keyhash(name string, logID [32]byte) uint32 {
+	h := sha256.New()
+	h.Write([]byte(name))
+	h.Write([]byte{0x0A, algRFC6962STH})
+	h.Write(logID[:])
+	r := h.Sum(nil)
+	return binary.BigEndian.Uint32(r)
+}
+
+// rfc6962LogName returns a sunlight checkpoint compatible log name from the
+// passed in CT log root URL.
+//
+// "For example, a log with submission prefix https://rome.ct.example.com/2024h1/ will use rome.ct.example.com/2024h1 as the checkpoint origin line"
```

Comment:
The quotes around this sentence have me asking a lot of questions. I suspect it's an accident. If not, it needs an attribution.

Reply:
Linked to Sunlight spec